### ISSUE TITLE
Cannot validate optional+null circulare fields

### ISSUE THREAD

**Issue by JakeShirley** [pelotom/runtypes#310]
I am seeing some odd behavior with circular types that are optional/nullable.  Perhaps I am constructing the record wrong though.

I'd like a property that is a nested type that can either be completely, missing, `null`, or defined.

Runtypes Version: 6.6.0

## Example Code
```ts
import { Boolean, Number, String, Literal, Array, Record, Union, Optional, Null, Lazy } from "runtypes";

export const Person = Lazy(() =>
  Record({
    name: String,
    partner: Optional(Person).Or(Null),
  })
);

try {
  const jsonBlob1 = JSON.parse(`{ "name": "Bob", "partner": { "name": "Janet" }}`);
  Person.check(jsonBlob1);
  console.info("Scenario 1 succeeded!");
} catch (e) {
  console.error(`Scenario 1 failed: ${e}`);
}

try {
  const jsonBlob2 = JSON.parse(`{ "name": "Bob", "partner": null }`);
  Person.check(jsonBlob2);
  console.info("Scenario 2 succeeded!");
} catch (e) {
  console.error(`Scenario 2 failed: ${e}`);
}

try {
  const jsonBlob3 = JSON.parse(`{ "name": "Bob" }`);
  Person.check(jsonBlob3);
  console.info("Scenario 3 succeeded!");
} catch (e) {
  console.error(`Scenario 3 failed: ${e}`);
}
```

## Output
```
Scenario 1 failed: ValidationError: Validation failed:
{
  "partner": "Expected { name: string; partner: CIRCULAR union; } | undefined | null, but was object"
}.
Object should match { name: string; partner: { name: string; partner: CIRCULAR union; } | undefined | null; }
Scenario 2 succeeded!
Scenario 3 failed: ValidationError: Validation failed:
{
  "partner": "Expected { name: string; partner: CIRCULAR union; } | undefined | null, but was missing"
}.
Object should match { name: string; partner: { name: string; partner: CIRCULAR union; } | undefined | null; }
```

**Comment by Deco** [comment]
Try switching the order of optional and Or Null, and let us know how it goes:
So instead of `Optional(Person).Or(Null),`, try `Optional(Person.Or(Null))`.

**Comment by Deco** [comment]
I just tried my own suggestion and it worked.

```
const {Boolean, Number, String, Literal, Array, Record, Union, Optional, Null, Lazy} = require("runtypes");

const Person = Lazy(() =>
    Record({
        name: String,
        partner: Optional(Person.Or(Null)),
    })
);

try {
    const jsonBlob1 = JSON.parse(`{ "name": "Bob", "partner": { "name": "Janet" }}`);
    Person.check(jsonBlob1);
    console.info("Scenario 1 succeeded!");
} catch (e) {
    console.error(`Scenario 1 failed: ${e}`);
}

try {
    const jsonBlob2 = JSON.parse(`{ "name": "Bob", "partner": null }`);
    Person.check(jsonBlob2);
    console.info("Scenario 2 succeeded!");
} catch (e) {
    console.error(`Scenario 2 failed: ${e}`);
}

try {
    const jsonBlob3 = JSON.parse(`{ "name": "Bob" }`);
    Person.check(jsonBlob3);
    console.info("Scenario 3 succeeded!");
} catch (e) {
    console.error(`Scenario 3 failed: ${e}`);
}
```

```
Scenario 1 succeeded!
Scenario 2 succeeded!
Scenario 3 succeeded!
```

The order of operations here is a bit of a gotcha. Perhaps even warning worthy?

**Comment by yuhr** [comment]
Yeah, if you want mark your property to be optional (i.e. able to be absent), `Optional` must be the outmost wrapper of your property. `Optional(runtype)` is no different from `Union(runtype, Undefined)` in any places other than direct children of a `Record`. If you wrote `partner: Optional(Person).Or(Null)` it's just equivalent to `partner: Union(Optional(Person), Null)` and then `partner: Union(Union(Person, Undefined), Null)`, so it means the property is not be able to be absent and be `Person | undefined | null`.

To summarize the correspondence of the runtypes to the TS's types:

- `Record({ property: Optional(String).Or(Null) })`: `{ property: string | null | undefined }`
- `Record({ property: Optional(String.Or(Null)) })`: `{ property?: string | null | undefined }`

**Comment by JakeShirley** [comment]
Excellent, thanks for the quick responses!  This makes total sense, if your example @yuhr isn't in the README/wiki, I'd suggest adding it.  It made it very clear!
